ludwig/models/retrieval.py

```diff
@@ -0,0 +1,224 @@
+import hashlib
+import json
+import os
+from typing import Any, Callable, Dict, List, Optional, Type, TYPE_CHECKING, Union
+
+import numpy as np
+import pandas as pd
+from tqdm import tqdm
+
+try:
+    import faiss
+except ImportError:
+    faiss = None
+
+if TYPE_CHECKING:
+    from ludwig.backend.base import Backend
+
+from ludwig.utils.batch_size_tuner import BatchSizeEvaluator
+from ludwig.utils.torch_utils import get_torch_device
+
+
+def df_checksum(df: pd.DataFrame) -> str:
+    return hashlib.sha1(pd.util.hash_pandas_object(df).values).hexdigest()
+
+
+def df_to_row_strs(df: pd.DataFrame) -> List[str]:
+    rows = df.to_dict(orient="records")
+    row_strs = [json.dumps(r) for r in rows]
+    return row_strs
+
+
+class RetrievalModel:
```

Comment:
Nit: use `ABC` as parent and decorate all methods as `@abstractmethod` with implementation `pass`.

Reply:
Done